lambda_function.py

```diff
@@ -293,64 +189,203 @@ def gather_table_stats(cursor, cluster):
+            else:
+                raise
+
+        output_metrics = []
+
+        for command in external_commands:
+            if command['type'] == 'value':
+                cmd_type = "Query"
+            else:
+                cmd_type = "Canary"
+
+            logger.info("Executing %s %s: %s" % (command_set_type, cmd_type, command['name']))
+
+            t = datetime.datetime.now()
+            interval = self.run_command(cursor, command['query'])
+            try:
+                value = cursor.fetchone()[0]
+            except pg8000.core.ProgrammingError:
+                logger.info("No values fetched for commamnd {}".format(command['name']))
+                value = None
+
+            if value is None:
+                value = 0
+
+            # append a cloudwatch metric for the value, or the elapsed interval, based upon the configured 'type' value
+            if command['type'] == 'value':
+                output_metrics.append({
+                    'MetricName': command['name'],
+                    'Dimensions': [
+                        {'Name': self.environment, 'Value': self.environment}
+                    ],
+                    'Timestamp': t,
+                    'Value': value,
+                    'Unit': command['unit']
+                })
+            else:
+                output_metrics.append({
+                    'MetricName': command['name'],
+                    'Dimensions': [
+                        {'Name': self.environment, 'Value': self.cluster}
+                    ],
+                    'Timestamp': t,
+                    'Value': interval,
+                    'Unit': 'Milliseconds'
+                })
+
+        return output_metrics
+
+    @staticmethod
+    def run_command(cursor, statement):
+        logger.debug("Running Statement: %s" % statement)
+
+        t = datetime.datetime.now()
+        try:
+            cursor.execute(statement)
+        except pg8000.core.ProgrammingError as e:
+            logger.error('Error executing statement %s: %s', statement, e)
+        interval = (datetime.datetime.now() - t).microseconds / 1000
+
+        return interval
+
+    def gather_service_class_stats(self, cursor):
+        metrics = []
+        poll_ts = datetime.datetime.utcnow()
+        _ = self.run_command(cursor, '''SELECT service_class, num_queued_queries, num_executing_queries 
+                                            FROM stv_wlm_service_class_state w WHERE w.service_class >= 6 ORDER BY 1
+                                    ''')
+        service_class_info = cursor.fetchall()
+
+        for service_class in service_class_info:
+            queued_metric = {'MetricName': 'ServiceClass%s-Queued' % service_class[0],
+                             'Dimensions': [{'Name': self.environment, 'Value': self.cluster}],
+                             'Timestamp': poll_ts,
+                             'Value': service_class[1]}
+            metrics.append(queued_metric.copy())
```

Comment:
Note -- `.copy()` won't deep copy -- so your ref to `service_class[1]` will remain. Look into `deepcopy` or even make a small recurse function:

```python
def deep_copy_primitives(value):
    if isinstance(value, dict):
        return {k: deep_copy_primitives(v) for k, v in value.iteritems()}
    elif isinstance(value, list):
        return [deep_copy_primitives(v) for v in in value]
    return value
```

Reply:
This is interesting, as it should result in interesting bugs in the original code, will swap for `deepcopy`, and see how/if they react .